openhcl/virt_mshv_vtl/src/processor/tdx/tlb_flush.rs

```diff
@@ -172,16 +185,98 @@ impl UhProcessor<'_, TdxBacked> {
             runner.invgla(gla_flags, gla_list).unwrap();
         };
 
-        *gva_list_count = partition_flush_state.s.gva_list_count;
+        *gva_list_count = partition_list_count;
         true
     }
 
     fn set_flush_entire(global: bool, vp_flags: &mut TdxVmFlags) {
         if global {
-            // TODO: Track EPT invalidations separately.
+            // TODO TDX: Track EPT invalidations separately.
             vp_flags.set_invd_translations(x86defs::tdx::TDX_VP_ENTER_INVD_INVEPT);
         } else if !global && vp_flags.invd_translations() == 0 {
             vp_flags.set_invd_translations(x86defs::tdx::TDX_VP_ENTER_INVD_INVVPID_NON_GLOBAL);
         }
     }
 }
+
+#[derive(Debug, Inspect)]
+pub(super) struct AtomicTlbRingBuffer {
+    /// The contents of the buffer.
+    #[inspect(hex, with = "|x| inspect::iter_by_index(x.iter())")]
+    buffer: Box<[AtomicU64; FLUSH_GVA_LIST_SIZE]>,
+    /// The number of GVAs that have been added over the lifetime of the VM.
+    gva_list_count: AtomicUsize,
+    /// The number of GVAs that have started being added to the list over the
+    /// lifetime of the VM.
+    in_progress_count: AtomicUsize,
+    /// A guard to ensure that only one thread is writing to the list at a time.
+    write_lock: Mutex<()>,
+}
+
+pub(super) struct AtomicTlbRingBufferWriteGuard<'a> {
+    buf: &'a AtomicTlbRingBuffer,
+    _write_lock: MutexGuard<'a, ()>,
+}
+
+impl AtomicTlbRingBuffer {
+    fn new() -> Self {
+        Self {
+            buffer: Box::new(std::array::from_fn(|_| AtomicU64::new(0))),
+            gva_list_count: AtomicUsize::new(0),
+            in_progress_count: AtomicUsize::new(0),
+            write_lock: Mutex::new(()),
+        }
+    }
+
+    fn count(&self) -> usize {
+        self.gva_list_count.load(Ordering::Acquire)
+    }
+
+    pub fn write(&self) -> AtomicTlbRingBufferWriteGuard<'_> {
+        let write_lock = self.write_lock.lock();
+        AtomicTlbRingBufferWriteGuard {
+            buf: self,
+            _write_lock: write_lock,
+        }
+    }
+
+    fn try_copy(&self, start_count: Wrapping<usize>, flush_addrs: &mut [u64]) -> bool {
+        let mut index = start_count;
+        for flush_addr in flush_addrs.iter_mut() {
+            *flush_addr = self.buffer[index.0 % FLUSH_GVA_LIST_SIZE].load(Ordering::Acquire);
```

Comment:
You can also use a relaxed load + a single atomic acquire fence after the loop. I expect the generated code to be the same on x86, so it may not be worth the trouble.